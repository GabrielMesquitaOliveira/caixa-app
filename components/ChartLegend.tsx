import { EstatisticasContratos } from '@/hooks/useHomeData';
import { formatarMoeda } from '@/utils/formatters';
import React from 'react';
import { Text, View } from 'react-native';

interface ChartLegendProps {
  estatisticas: EstatisticasContratos;
}

export const ChartLegend: React.FC<ChartLegendProps> = ({ estatisticas }) => {
  const dadosLegenda = [
    { name: "Pago", value: estatisticas.totalPago, color: "#22c55e" },
    { name: "Pendente", value: estatisticas.totalPendente, color: "#f59e0b" },
    { name: "Saldo Devedor", value: estatisticas.saldoTotalDevedor, color: "#ef4444" }
  ];

  return (
    <View className="w-11/12 mx-auto mb-8">
      <Text className="text-2xl font-semibold text-gray-800 mb-3">Distribuição Financeira</Text>
      <View className="bg-white rounded-xl p-4 shadow-md">
        {dadosLegenda.map((item, index) => (
          <View key={index} className="flex flex-row items-center justify-between py-2">
            <View className="flex flex-row items-center">
              <View 
                className="w-4 h-4 rounded-full mr-3"
                style={{ backgroundColor: item.color }}
              />
              <Text className="text-sm font-medium text-gray-700">{item.name}</Text>
            </View>
            <Text className="text-sm font-semibold text-gray-800">
              {formatarMoeda(item.value)}
            </Text>
          </View>
        ))}
      </View>
    </View>
  );
};