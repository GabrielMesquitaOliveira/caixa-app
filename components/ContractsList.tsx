import { Contrato, Parcela } from '@/services';
import { formatarMoeda } from '@/utils/formatters';
import React from 'react';
import { Text, View } from 'react-native';

interface ContractsListProps {
  contratos: Contrato[];
  parcelas: Parcela[];
}

export const ContractsList: React.FC<ContractsListProps> = ({ contratos, parcelas }) => {
  return (
    <View className="w-11/12 mx-auto mt-6 mb-8">
      <Text className="text-2xl font-semibold text-gray-800 mb-4">Seus Empr√©stimos</Text>
      
      {contratos.map((contrato) => {
        const parcelasContrato = parcelas.filter(p => p.contratoId === contrato.id);
        const parcelasPagas = parcelasContrato.filter(p => p.situacao === 'paga').length;
        const percentualContrato = parcelasContrato.length > 0 
          ? (parcelasPagas / parcelasContrato.length) * 100 
          : 0;

        return (
          <View key={contrato.id} className="bg-white rounded-xl p-4 shadow-md mb-3">
            <View className="flex flex-row justify-between items-start">
              <View className="flex-1 w-2/3">
                <Text className="text-lg font-semibold text-gray-800">
                  {contrato.nomePersonalizado || contrato.produto?.nome}
                </Text>
                <Text className="text-sm text-gray-600 mt-1">
                  Contrato: {contrato.numeroContrato}
                </Text>
                <Text className="text-sm text-gray-600">
                  Sistema: {contrato.sistemaAmortizacao || 'PRICE'}
                </Text>
              </View>
              <View className="items-end w-1/3">
                <Text className="text-lg font-bold text-blue-600">
                  {formatarMoeda(contrato.valorContratado)}
                </Text>
                <Text className="text-xs text-gray-500">
                  {parcelasPagas}/{contrato.prazoContratado || parcelasContrato.length} parcelas
                </Text>
              </View>
            </View>
            
            <View className="mt-3">
              <View className="w-full bg-gray-200 rounded-full h-2">
                <View 
                  className="bg-blue-600 h-2 rounded-full" 
                  style={{ width: `${percentualContrato}%` }}
                />
              </View>
              <Text className="text-xs text-gray-500 mt-1">
                {Math.round(percentualContrato)}% pago
              </Text>
            </View>
          </View>
        );
      })}
    </View>
  );
};
