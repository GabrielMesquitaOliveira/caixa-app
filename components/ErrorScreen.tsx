import React from 'react';
import { Text, View } from 'react-native';

interface ErrorScreenProps {
  error: string;
}

export const ErrorScreen: React.FC<ErrorScreenProps> = ({ error }) => (
  <View className="h-full w-full flex items-center justify-center bg-neutral-200">
    <Text className="text-lg text-red-600">{error}</Text>
  </View>
);