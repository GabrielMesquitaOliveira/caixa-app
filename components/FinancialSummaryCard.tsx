import { EstatisticasContratos } from '@/hooks/useHomeData';
import { formatarMoeda } from '@/utils/formatters';
import React from 'react';
import { Text, View } from 'react-native';
import { PieChart } from 'react-native-chart-kit';

interface FinancialSummaryCardProps {
  estatisticas: EstatisticasContratos;
}

/**
 * Componente que exibe um resumo financeiro com dados de contratos ativos,
 * saldo devedor total e uma grafico de pizza com a distribui o dos valores
 * pagos, pendentes e saldo devedor.
 * 
 * @example
 * <FinancialSummaryCard estatisticas={estatisticas} />
 * 
 * @param estatisticas - objeto com as estat√≠sticas de contratos ativos
 * @returns JSX.Element
 */
export const FinancialSummaryCard: React.FC<FinancialSummaryCardProps> = ({ estatisticas }) => {
  const dadosGrafico = [
    {
      name: "Pago",
      population: estatisticas.totalPago,
      color: "#22c55e",
      legendFontColor: "#333",
      legendFontSize: 12
    },
    {
      name: "Pendente",
      population: estatisticas.totalPendente,
      color: "#f59e0b",
      legendFontColor: "#333",
      legendFontSize: 12
    },
    {
      name: "Saldo Devedor",
      population: estatisticas.saldoTotalDevedor,
      color: "#ef4444",
      legendFontColor: "#333",
      legendFontSize: 12
    }
  ];

  const chartConfig = {
    backgroundColor: "#ffffff",
    backgroundGradientFrom: "#ffffff",
    backgroundGradientTo: "#ffffff",
    color: (opacity = 1) => `rgba(0, 0, 0, ${opacity})`,
    labelColor: (opacity = 1) => `rgba(0, 0, 0, ${opacity})`,
  };

  return (
    <View className="w-11/12 h-60 mx-auto rounded-2xl bg-white flex flex-row items-center justify-center -mt-20 p-6 shadow-lg">
      <View className="w-2/3 flex flex-col justify-center">
        <Text className="text-xl font-semibold text-gray-700">Saldo Devedor</Text>
        <Text className="text-3xl font-semibold text-red-600">
          {formatarMoeda(estatisticas.saldoTotalDevedor)}
        </Text>
        <Text className="text-sm text-gray-500 mt-1">
          {estatisticas.totalContratos} contrato(s) ativo(s)
        </Text>
      </View>
      <View className="w-1/3 flex flex-col h-full justify-center items-center">
        {dadosGrafico.length > 0 && (
          <PieChart
            data={dadosGrafico}
            width={120}
            height={120}
            chartConfig={chartConfig}
            accessor="population"
            backgroundColor="transparent"
            paddingLeft="0"
            absolute={false}
            hasLegend={false}
          />
        )}
      </View>
    </View>
  );
};
