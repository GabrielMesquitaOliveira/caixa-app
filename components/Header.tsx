import MaterialIcons from '@expo/vector-icons/MaterialIcons';
import type { NativeStackHeaderProps } from '@react-navigation/native-stack';
import React from 'react';
import { Text, TouchableOpacity, View } from 'react-native';
import { SafeAreaView } from 'react-native-safe-area-context';

// Estendendo o tipo das options para incluir props customizadas
interface CustomHeaderOptions {
  numeroConta?: string;
  onMenuPress?: () => void;
  onNotificacoesPress?: () => void;
  showAccountNumber?: boolean;
}

interface CustomHeaderProps extends NativeStackHeaderProps {
  options: NativeStackHeaderProps['options'] & CustomHeaderOptions;
}

export function Header({ 
  navigation, 
  route, 
  options,
  back,
}: CustomHeaderProps) {
  // Props padrão ou vindas das options da screen
  const numeroConta = (options as any).numeroConta || "Ag. 3737 CC.576939466-4";
  const showAccountNumber = (options as any).showAccountNumber ?? !back;
  const title = options.title || route.name;

  const handleBackPress = () => {
    if (navigation.canGoBack()) {
      navigation.goBack();
    }
  };

  const handleMenuPress = () => {
    const onMenuPress = (options as any).onMenuPress;
    if (onMenuPress) {
      onMenuPress();
    } else {
      // Comportamento padrão do menu
      console.log('Menu pressed');
    }
  };

  const handleNotificacoesPress = () => {
    const onNotificacoesPress = (options as any).onNotificacoesPress;
    if (onNotificacoesPress) {
      onNotificacoesPress();
    } else {
      // Comportamento padrão das notificações
      console.log('Notificações pressed');
    }
  };

  return (
    <SafeAreaView edges={['top']} className="bg-[#005CA9]">
      <View className="flex-row items-center justify-between px-4 py-3 bg-[#005CA9]">
        {/* Left Side - Menu ou Back Button */}
        <TouchableOpacity 
          onPress={back ? handleBackPress : handleMenuPress} 
          className="p-2"
        >
          <MaterialIcons
            name={back ? "arrow-back" : "menu"}
            size={24}
            color="#fff"
          />
        </TouchableOpacity>

        {/* Center - Numero da conta ou título */}
        <View className="flex-1 items-center justify-center">
          <Text className="text-white text-2xl font-bold">
            {showAccountNumber ? numeroConta : title}
          </Text>
        </View>

        {/* Right Side - Notificações */}
        <TouchableOpacity onPress={handleNotificacoesPress} className="p-2">
          <MaterialIcons
            name="notifications"
            size={24}
            color="#fff"
          />
        </TouchableOpacity>
      </View>
    </SafeAreaView>
  );
}