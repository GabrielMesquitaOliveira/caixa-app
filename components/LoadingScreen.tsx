import React from 'react';
import { ActivityIndicator, Text, View } from 'react-native';

export const LoadingScreen: React.FC = () => (
  <View className="h-full w-full flex items-center justify-center bg-neutral-200">
    <ActivityIndicator size="large" color="#005CA9" />
    <Text className="text-lg text-gray-600 mt-4">Carregando...</Text>
  </View>
);