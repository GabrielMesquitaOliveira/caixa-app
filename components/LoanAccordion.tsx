import {
  Accordion,
  AccordionContent,
  AccordionContentText,
  AccordionHeader,
  AccordionIcon,
  AccordionItem,
  AccordionTitleText,
  AccordionTrigger
} from '@/components/ui/accordion';
import { ChevronDownIcon } from '@/components/ui/icon';
import { formatarMoeda, formatarPorcentagem } from '@/utils/formatters';
import React from 'react';
import { View } from 'react-native';

interface LoanAccordionProps {
  contrato: {
    valorContratado: number;
    dataContratacao: string;
    sistemaAmortizacao: string;
    taxaJurosMensal: number;
    taxaJurosAnual: number;
    prazoContratado: number;
    parcelasPagas: number;
    saldoDevedor: number;
    dataUltimoVencimento: string;
    cetMensal?: number;
    cetAnual?: number;
  };
}

const LoanAccordion: React.FC<LoanAccordionProps> = ({ contrato }) => {
  const prazoRemanescente = contrato.prazoContratado - contrato.parcelasPagas;
  
  return (
    <Accordion variant="filled" size="md" className="mb-3">
      <AccordionItem value="details">
        <AccordionHeader>
          <AccordionTrigger>
            <AccordionTitleText>Detalhes do Empréstimo</AccordionTitleText>
            <AccordionIcon as={ChevronDownIcon} />
          </AccordionTrigger>
        </AccordionHeader>
        <AccordionContent>
          <View className="space-y-2">
            <View className="flex-row justify-between">
              <AccordionContentText className="text-gray-600">Valor Contratado:</AccordionContentText>
              <AccordionContentText className="font-semibold">{formatarMoeda(contrato.valorContratado)}</AccordionContentText>
            </View>
            
            <View className="flex-row justify-between">
              <AccordionContentText className="text-gray-600">Data de Contratação:</AccordionContentText>
              <AccordionContentText className="font-semibold">
                {new Date(contrato.dataContratacao).toLocaleDateString('pt-BR')}
              </AccordionContentText>
            </View>
            
            <View className="flex-row justify-between">
              <AccordionContentText className="text-gray-600">Sistema de Amortização:</AccordionContentText>
              <AccordionContentText className="font-semibold">{contrato.sistemaAmortizacao}</AccordionContentText>
            </View>
            
            <View className="flex-row justify-between">
              <AccordionContentText className="text-gray-600">Taxa de Juros Mensal:</AccordionContentText>
              <AccordionContentText className="font-semibold">{formatarPorcentagem(contrato.taxaJurosMensal)}</AccordionContentText>
            </View>
            
            <View className="flex-row justify-between">
              <AccordionContentText className="text-gray-600">Taxa de Juros Anual:</AccordionContentText>
              <AccordionContentText className="font-semibold">{formatarPorcentagem(contrato.taxaJurosAnual)}</AccordionContentText>
            </View>
            
            <View className="flex-row justify-between">
              <AccordionContentText className="text-gray-600">Prazo Total (Meses):</AccordionContentText>
              <AccordionContentText className="font-semibold">{contrato.prazoContratado}</AccordionContentText>
            </View>
            
            <View className="flex-row justify-between">
              <AccordionContentText className="text-gray-600">Prazo Remanescente:</AccordionContentText>
              <AccordionContentText className="font-semibold">{prazoRemanescente} meses</AccordionContentText>
            </View>
            
            <View className="flex-row justify-between">
              <AccordionContentText className="text-gray-600">Saldo Devedor Atualizado:</AccordionContentText>
              <AccordionContentText className="font-semibold">{formatarMoeda(contrato.saldoDevedor)}</AccordionContentText>
            </View>
            
            <View className="flex-row justify-between">
              <AccordionContentText className="text-gray-600">Data Último Vencimento:</AccordionContentText>
                <AccordionContentText className="font-semibold">
                {new Date(contrato.dataUltimoVencimento).toLocaleDateString('pt-BR')}
                </AccordionContentText>
            </View>
            
            {contrato.cetMensal && (
              <View className="flex-row justify-between">
                <AccordionContentText className="text-gray-600">CET Mensal:</AccordionContentText>
                <AccordionContentText className="font-semibold">{formatarPorcentagem(contrato.cetMensal)}</AccordionContentText>
              </View>
            )}
            
            {contrato.cetAnual && (
              <View className="flex-row justify-between">
                <AccordionContentText className="text-gray-600">CET Anual:</AccordionContentText>
                <AccordionContentText className="font-semibold">{formatarPorcentagem(contrato.cetAnual)}</AccordionContentText>
              </View>
            )}
          </View>
        </AccordionContent>
      </AccordionItem>
    </Accordion>
  );
};

export default LoanAccordion;
