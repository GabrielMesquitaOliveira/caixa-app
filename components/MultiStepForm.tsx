import { FormData } from '@/types';
import { yupResolver } from '@hookform/resolvers/yup';
import React, { useState } from 'react';
import { useForm } from 'react-hook-form';
import { Text, View } from 'react-native';
import * as yup from 'yup';
import StepOne from './StepOne';
import StepThree from './StepThree';
import StepTwo from './StepTwo';

interface MultiStepFormProps {
  onSubmit: (data: FormData) => void;
  onCancel: () => void;
}

// Schema de validação para cada passo
const validationSchemas = {
  stepOne: yup.object().shape({
    valorContratado: yup
      .number()
      .required('Valor é obrigatório')
      .min(1000, 'Valor mínimo é R$ 1.000')
      .max(1000000, 'Valor máximo é R$ 1.000.000'),
    prazoContratado: yup
      .number()
      .required('Prazo é obrigatório')
      .min(1, 'Prazo mínimo é 1 mês')
      .max(360, 'Prazo máximo é 360 meses'),
  }),
  stepTwo: yup.object().shape({
    sistemaAmortizacao: yup
      .string()
      .required('Sistema de amortização é obrigatório')
      .oneOf(['PRICE', 'SAC'], 'Selecione um sistema válido'),
    dataVencimento: yup
      .string()
      .required('Data de vencimento é obrigatória'),
  }),
  stepThree: yup.object().shape({
    nomePersonalizado: yup
      .string()
      .required('Nome personalizado é obrigatório')
      .min(3, 'Mínimo 3 caracteres')
      .max(50, 'Máximo 50 caracteres'),
    observacoes: yup
      .string()
      .max(500, 'Máximo 500 caracteres'),
  }),
};

const MultiStepForm: React.FC<MultiStepFormProps> = ({ onSubmit, onCancel }) => {
  const [step, setStep] = useState(1);
  const [formData, setFormData] = useState<FormData>({});

  // Hook forms separados para cada passo
  const stepOneForm = useForm({
    resolver: yupResolver(validationSchemas.stepOne),
  });

  const stepTwoForm = useForm({
    resolver: yupResolver(validationSchemas.stepTwo),
  });

  const stepThreeForm = useForm({
    resolver: yupResolver(validationSchemas.stepThree),
  });

  const onNextStep = (data: Partial<FormData>) => {
    setFormData({ ...formData, ...data });
    if (step < 3) {
      setStep(step + 1);
    } else {
      onSubmit({ ...formData, ...data } as FormData);
    }
  };

  const onPrevious = () => {
    if (step > 1) {
      setStep(step - 1);
    }
  };

  return (
    <View className="p-4">
      {/* Indicador de progresso */}
      <View className="flex-row justify-center mb-6">
        {[1, 2, 3].map((stepNumber) => (
          <View
            key={stepNumber}
            className={`w-8 h-8 rounded-full mx-2 flex items-center justify-center ${
              stepNumber <= step ? 'bg-blue-500' : 'bg-gray-300'
            }`}
          >
            <Text className="text-white font-bold">{stepNumber}</Text>
          </View>
        ))}
      </View>

      {/* Formulário atual */}
      {step === 1 && (
        <StepOne
          control={stepOneForm.control as any}
          errors={stepOneForm.formState.errors}
          handleSubmit={stepOneForm.handleSubmit}
          onNextStep={onNextStep}
          onCancel={onCancel}
        />
      )}
      {step === 2 && (
        <StepTwo
          control={stepTwoForm.control as any}
          errors={stepTwoForm.formState.errors}
          handleSubmit={stepTwoForm.handleSubmit}
          onNextStep={onNextStep}
          onPrevious={onPrevious}
          onCancel={onCancel}
          formData={formData}
        />
      )}
      {step === 3 && (
        <StepThree
          control={stepThreeForm.control as any}
          errors={stepThreeForm.formState.errors}
          handleSubmit={stepThreeForm.handleSubmit}
          onNextStep={onNextStep}
          onPrevious={onPrevious}
          onCancel={onCancel}
          formData={formData}
        />
      )}
    </View>
  );
};

export default MultiStepForm;
