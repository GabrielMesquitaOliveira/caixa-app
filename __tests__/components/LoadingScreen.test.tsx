import { LoadingScreen } from '@/components/LoadingScreen';
import { render } from '@testing-library/react-native';
import React from 'react';

describe('LoadingScreen Component', () => {
  it('renders correctly', () => {
    const { getByText } = render(<LoadingScreen />);
    // Add assertions based on the expected output
    expect(getByText('Expected Text')).toBeTruthy(); // Replace 'Expected Text' with actual text
  });

  it('matches snapshot', () => {
    const { toJSON } = render(<LoadingScreen />);
    expect(toJSON()).toMatchSnapshot();
  });
});
