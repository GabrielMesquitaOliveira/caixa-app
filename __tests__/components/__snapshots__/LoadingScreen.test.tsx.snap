// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoadingScreen Component matches snapshot 1`] = `
<View
  className="h-full w-full flex items-center justify-center bg-neutral-200"
>
  <ActivityIndicator
    color="#005CA9"
    size="large"
  />
  <Text
    className="text-lg text-gray-600 mt-4"
  >
    Carregando...
  </Text>
</View>
`;
